build: use modern cmake

dundargoc opened the pull request:
Replace old-school cmake with the so-called "Modern CMake", meaning
preferring using targets and properties over directory settings and
variables. This allows greater flexibility, robustness and clarity over
how the code works.

The following deprecated commands will be replaced with their modern
alternatives that operates on a specific target, rather than all targets
in the current directory:

- add_compile_options -> target_compile_options
- include_directories -> target_include_directories
- link_libraries -> target_link_libraries
- add_definitions -> target_compile_definitions

There are mainly four main targets that we currently use: nvim, libnvim,
nvim-test (used by unittests) and ${texe} (used by
check-single-includes). The goal is to explicitly define the
dependencies of each target fully, rather than having everything be
dependent on everything else.

theHamsta commented on src/nvim/CMakeLists.txt:
```diff
@@ -446,6 +446,108 @@ add_executable(nvim ${NVIM_GENERATED_FOR_SOURCES} ${NVIM_GENERATED_FOR_HEADERS}
+    link_libraries(-fstack-protector-strong)
+  elseif(HAS_FSTACK_PROTECTOR_FLAG)
+    target_compile_options(nvim PRIVATE -fstack-protector --param ssp-buffer-size=4)
+    link_libraries(-fstack-protector --param ssp-buffer-size=4)
```
target_link_libraries?

dundargoc replied:
Yeah, working on it :)

justinmk commented:
nice. does this change the min required cmake version ?

would you mind giving a small concrete example in the commit message of "X used to be like ..., now it's done like ...". Not obvious from the diff what's going on.

dundargoc commented:
> nice. does this change the min required cmake version ?

No. "Modern" cmake is almost 10 years old now.

> would you mind giving a small concrete example in the commit message of "X used to be like ..., now it's done like ...". Not obvious from the diff what's going on.

Sure.

ii14 commented on CMakeLists.txt:
```diff
@@ -541,20 +324,29 @@ endif()
+  ${LIBTERMKEY_INCLUDE_DIRS}
+  ${LIBUV_INCLUDE_DIRS}
+  ${LIBVTERM_INCLUDE_DIRS}
+  ${LibIntl_INCLUDE_DIRS}
```
`${LibIntl_INCLUDE_DIRS}` will be defined only with `ENABLE_LIBINTL` (same with `${Iconv_INCLUDE_DIRS}`). Is including non-existent variables okay?

dundargoc replied:
Should be fine since non-existing variables evaluate to an empty string. However, maybe it's clearer to do

```cmake
if(ENABLE_LIBINTL)
  # LibIntl (not Intl) selects our FindLibIntl.cmake script. #8464
  find_package(LibIntl REQUIRED)
  target_include_directories(main_lib SYSTEM INTERFACE ${LibIntl_INCLUDE_DIRS})
  target_link_libraries(main_lib INTERFACE ${LibIntl_LIBRARY})
endif()
```

directly, rather than collect them in a variable?

ii14 commented on src/nvim/CMakeLists.txt:
```diff
@@ -388,80 +501,246 @@ foreach(hfile ${NVIM_GENERATED_FOR_HEADERS})
-set(NVIM_EXEC_LINK_LIBRARIES ${NVIM_LINK_LIBRARIES} ${LUA_PREFERRED_LIBRARIES})
+add_library(main_lib INTERFACE)
+target_link_libraries(main_lib INTERFACE
+  definition_lib
+  include_lib
+  link_lib
+  option_lib)
```
If these are always used together, it can be one single library

dundargoc replied:
I was thinking of `texe` on line 938, but I think we should be able to add these libraries as well. On it.

dundargoc committed:
build: use modern cmake

Replace old-school cmake with the so-called "Modern CMake", meaning
preferring using targets and properties over directory settings and
variables. This allows greater flexibility, robustness and clarity over
how the code works.

The following deprecated commands will be replaced with their modern
alternatives that operates on a specific target, rather than all targets
in the current directory:

- add_compile_options -> target_compile_options
- include_directories -> target_include_directories
- link_libraries -> target_link_libraries
- add_definitions -> target_compile_definitions

There are mainly four main targets that we currently use: nvim, libnvim,
nvim-test (used by unittests) and ${texe} (used by
check-single-includes). The goal is to explicitly define the
dependencies of each target fully, rather than having everything be
dependent on everything else.

cryptomilk commented on src/nvim/CMakeLists.txt:
```diff
@@ -639,17 +639,26 @@ if(CLANG_ASAN_UBSAN)
+    -fno-optimize-sibling-calls
+    -fsanitize=address
+    -fsanitize=undefined
+    -fsanitize-blacklist=${PROJECT_SOURCE_DIR}/src/.asan-blacklist)
```
I would suggest to use `${nvim_SOURCE_DIR}`. You never know if the project is included by a another one and this would point to the wrong one :-)

dundargoc replied:
Hmmm, wouldn't the same dilemma exist for the parent project? They wouldn't be able to use `PROJECT_SOURCE_DIR` in case another project includes *their* project :)

That being said: I don't mind this, so I'll include it 👍

dundargoc replied:
Actually in hindsight, the mixing of lowercase and uppercase is particularly fugly, even for cmake standards. Is there an external project this currently breaks?

cryptomilk replied:
I have no idea. I learned it the hard way through bug reports for two of my projects (cmocka, libssh).

dundargoc replied:
Honestly it's too ugly lol. I'm willing to risk it. You can tell me "I told you so" if this ends up biting me in the ass later on :)

dundargoc replied:
Also this is a relatively unknown feature, so it's gonna be annoying for new contributors to figure out where the variable came from.

dundargoc commented on src/nvim/CMakeLists.txt:
```diff
@@ -631,29 +899,37 @@ if(CLANG_ASAN_UBSAN)
     if(CI_BUILD)
       # Try to recover from all sanitize issues so we get reports about all failures
-      set(SANITIZE_RECOVER -fsanitize-recover)         # Clang 3.5-
+      target_compile_options(nvim PRIVATE -fsanitize-recover) # Clang 3.5-
```
clang version 3.5-? I wonder if neovim is even buildable with a clang that old. Should we remove this?

dundargoc replied:
I've already removed this in https://github.com/neovim/neovim/pull/21678 as I wanted to remove code outside of this PR so they wouldn't get lost in a massive squash.

It's not adding much cost but it's not adding much benefit either. This is a dev dependency/requirement, so I think we can afford to be less conservative in what we decide to remove. We can revert it if you'd like.

ii14 commented on src/nvim/CMakeLists.txt:
```diff
@@ -1,25 +1,349 @@
+
+if(CMAKE_SYSTEM_NAME STREQUAL "Darwin" AND CMAKE_SIZEOF_VOID_P EQUAL 8 AND NOT PREFER_LUA AND LUAJIT_VERSION LESS "2.1.0-beta3")
+  # Required for luajit < 2.1.0-beta3.
+  target_link_libraries(nvim INTERFACE "-pagezero_size 10000" "-image_base 100000000")
+  target_link_libraries(nvim-test INTERFACE "-image_base 100000000")
```
I think in all those places you have to replace `INTERFACE` now. Not sure if this is relevant to that build failure on CI
```suggestion
  target_link_libraries(nvim PRIVATE "-pagezero_size 10000" "-image_base 100000000")
  target_link_libraries(nvim-test PRIVATE "-image_base 100000000")
```

dundargoc commented:
I'd like to squash before merging :)

justinmk commented on cmake.config/CMakeLists.txt:
```diff
@@ -5,6 +5,26 @@ include(CheckIncludeFiles)
 include(CheckCSourceRuns)
 include(CheckCSourceCompiles)
 
+check_c_source_compiles("
```
nice. is it feasible to mention in a comment at the top of cmakelists.txt, something like

> these things (`check_c_source_compiles `, foo, bar, ...) should be done in the *config* phase (cmake.config/), these other things are done in ...)

or is it too situation-dependent ?

justinmk replied:
oh i see this wouldn't make sense for subproject-specific checks (treesitter etc).

justinmk reviewed (approved):
the PR description is very nice, please include it in the commit message!
